sdk/nodejs/runtime/closure/v8_v11andHigher.ts

```diff
@@ -0,0 +1,214 @@
+    const functionId = await getRuntimeIdForFunctionAsync(func);
+
+    // Now, query for the internal properties the runtime sets up for it.
+    const { internalProperties } = await runtimeGetPropertiesAsync(functionId, /*ownProperties:*/ false);
+
+    // There should normally be an internal property called [[Scopes]]:
+    // https://chromium.googlesource.com/v8/v8.git/+/3f99afc93c9ba1ba5df19f123b93cc3079893c9b/src/inspector/v8-debugger.cc#820
+    const scopes = internalProperties.find(p => p.name === "[[Scopes]]");
+    if (!scopes) {
+        throw new Error("Could not find [[Scopes]] property");
+    }
+
+    if (!scopes.value) {
+        throw new Error("[[Scopes]] property did not have [value]");
+    }
+
+    if (!scopes.value.objectId) {
+        throw new Error("[[Scopes]].value have objectId");
+    }
+
+    // This is sneaky, but we can actually map back from the [[Scopes]] object to a real in-memory
+    // v8 array-like value.  Note: this isn't actually a real array.  For example, it cannot be
+    // iterated.  Nor can any actual methods be called on it. However, we can directly index into
+    // it, and we can.  Similarly, the 'object' type it optionally points at is not a true JS
+    // object.  So we can't call things like .hasOwnProperty on it.  However, the values pointed to
+    // by 'object' are the real in-memory JS objects we are looking for.  So we can find and return
+    // those successfully to our caller.
+    const scopesArray: { object?: Record<string, any> }[] = await getValueForObjectId(scopes.value.objectId);
+
+    // scopesArray is ordered from innermost to outermost.
+    for (let i = 0, n = scopesArray.length; i < n; i++) {
+        const scope = scopesArray[i];
+        if (scope.object) {
+            if (freeVariable in scope.object) {
+                const val = scope.object[freeVariable];
+                return val;
+            }
+        }
+    }
+
+    if (throwOnFailure) {
+        throw new Error("Unexpected missing variable in closure environment: " + freeVariable);
+    }
+
+    return undefined;
+}
+
+// In order to get information about an object, we need to put it in a well known location so that
+// we can call Runtime.evaluate and find it.  To do this, we just make a special map on the 'global'
+// object, and map from a unique-id to that object.  We then call Runtime.evaluate with an
+// expression that then points to that unique-id in that global object.  The runtime will then find
+// the object and give us back an internal id for it.  We can then query for information about the
+// object through that internal id.
+//
+// Note: the reason for the mapping object and the unique-id we create is so that we don't run into
+// any issues when being called asynchronously.  We don't want to place the object in a location
+// that might be overwritten by another call while we're asynchronously waiting for our original
+// call to complete.
+
+let currentFunctionId = 0;
```

Comment:
can we put this on `global` too? We have gotten burned time and time again by global state in `pulumi/pulumi`.